### PR TITLE
GEODE-5597 Publish geode artifacts to maven repo on GCS

### PR DESCRIPTION
**Pull request by FSOUTHERLAND**
Uses vanilla maven-publish Gradle plugin, instead of nexus (nexus is
still used for release artifact publishing)
GCS credentials are inherited from the concourse worker

Co-authored-by: Finn Southerland <fsoutherland@pivotal.io>
Co-authored-by: Jake Barrett <bjarrett@pivotal.io>

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by FSOUTHERLAND** [comment]
@pivotal-jbarrett @smgoller @PurelyApplied

**Comment by metatype** [comment]
As a followup to this PR I'd like to see us converge on a single maven publishing plugin.  Originally the standard gradle plugin didn't support signing pom files.  We need that in order to publish release artifacts so that's why we used nexus.  Hopefully that's been fixed and we can converge on the newer one.

**Comment by metatype** [comment]
Did the pom customizations in the `modifyPom` configuration of `publish.gradle` get applied?

**Comment by FSOUTHERLAND** [comment]
I believe 'modifyPom' is a nexus plugin thing only.

**Comment by metatype** [comment]
Have you compared the pom generated by nexus vs the maven plugin?  Are the dependencies exactly the same (including optional)?

**Comment by FSOUTHERLAND** [comment]
They are definitely not. Good catch.

**Comment by FSOUTHERLAND** [comment]
It looks like the pom files generated for nexus are actually incorrect (they include test dependencies), while the POM files generated by the maven publish plugin should now be correct.

**Comment by metatype** [comment]
Can you post the diff between the 2 poms?  That would make it easier to review if the changes are correct.

**Comment by FSOUTHERLAND** [comment]
[diff.txt](https://github.com/apache/geode/files/2303485/diff.txt)

Here is the diff of the POM files for geode-core

**Comment by FSOUTHERLAND** [comment]
[diff-depends.txt](https://github.com/apache/geode/files/2303645/diff-depends.txt)
Here is a diff of just the dependencies and their scope.

**Comment by metatype** [comment]
I'm fine with removing dependencies with scope `test` from the pom.  We need to preserve the `optional` element though.

**Comment by pivotal-jbarrett** [comment]
@metatype since we lack a really good external test its hard to tell if the pom changes are going to be an issue downstream. One change that seems odd is that changing dependencies from runtime to runtimeOnly resulted in a change of dependent jars as seen by the integration test that tests jars. Do you know if we really need to be shipping with dependencies on commons-codec and lucene-analyzers-phonetic? See https://concourse.apachegeode-ci.info/builds/19626 for the failure.

**Comment by pivotal-jbarrett** [comment]
@metatype this should be ready to roll. The distributed tests are failing for out of memory again.

**Comment by metatype** [comment]
Yep, the dependencies on commons-codec and lucene-analyzers-phonetic are required and should be on the runtime class path.

It makes sense that changing to a runtimeOnly dependency would break this (it's the equivalent of provided).

File: geode-assembly/src/integrationTest/resources/expected_jars.txt [hunk 1]
```diff
@@ -5,7 +5,6 @@ byte-buddy
 classgraph
 classmate
 commons-beanutils
-commons-codec
```
Comment (metatype) [inline geode-assembly/src/integrationTest/resources/expected_jars.txt]: Why is this removed?

File: geode-assembly/src/integrationTest/resources/expected_jars.txt [hunk 2]
```diff
@@ -47,7 +46,6 @@ log4j-jcl
 log4j-jul
 log4j-slf4j-impl
 lucene-analyzers-common
-lucene-analyzers-phonetic
```
Comment (metatype) [inline geode-assembly/src/integrationTest/resources/expected_jars.txt]: This shouldn't be removed.

File: geode-core/build.gradle [hunk 3]
```diff
@@ -86,19 +86,19 @@ dependencies {
 
   compile 'org.apache.logging.log4j:log4j-api:' + project.'log4j.version'
   compile 'org.apache.logging.log4j:log4j-core:' + project.'log4j.version'
-  runtime ('org.fusesource.jansi:jansi:' + project.'jansi.version') {
```
Comment (metatype) [inline geode-core/build.gradle]: Why is this changed?  This should be on the class path.

File: gradle/test.gradle [hunk 4]
```diff
@@ -117,6 +117,14 @@ subprojects {
     }
   }
 
+  // Facets does not extend the new runtimeOnly configurations
+  configurations.integrationTestRuntimeOnly.extendsFrom(configurations.runtimeOnly)
```
Comment (metatype) [inline gradle/test.gradle]: Can you do a PR to fix facets?

**Comment by FSOUTHERLAND** [comment]
@metatype changing deps to runtimeOnly is inline with the upcoming Gradle 5.0 changes and results in the maven POM showing the deps correctly as runtime scoped. The maven plugin shows 'runtime' dependencies as compile scoped otherwise.

File: gradle/test.gradle (hunk 4 above)
Comment (FSOUTHERLAND) [reply gradle/test.gradle]: You mean a PR to fix the facets plugin at the source? Looking at that the issue has been open for a year and many PRs stay stale for months. A PR to facets would have to be paired with this workaround.

**Comment by metatype** [comment]
Thanks for clearing up my misconception around `runtimeOnly`.

File: gradle/test.gradle (hunk 4 above)
Comment (metatype) [reply gradle/test.gradle]: It would be nice to improve the upstream plugins and keep workarounds out of our builds.  Oh well.

File: geode-core/build.gradle (hunk 3 above)
Comment (FSOUTHERLAND) [reply geode-core/build.gradle]: It is, but now as a runtimeOnly not runtime (deprecated).

**Comment by metatype** [comment]
Unfortunately the set of geode jars in the lib/ dir seem wrong when compared to our previous release (not the versions, but missing / incomplete).  I would also compare the manifest class path.

**Comment by pivotal-jbarrett** [comment]
I think we will write some new tests for asserting the more strictly the contents of the assembly that will require some explicit intervention to allow new or removed files.

**Comment by pivotal-jbarrett** [comment]
@metatype Merging #2359 confirms your assessment that geode jars are now missing.

```
    Removed Content
    --------------
    lib/geode-connectors-1.8.0-SNAPSHOT.jar
    lib/geode-cq-1.8.0-SNAPSHOT.jar
    lib/geode-lucene-1.8.0-SNAPSHOT.jar
    lib/geode-old-client-support-1.8.0-SNAPSHOT.jar
    lib/geode-protobuf-1.8.0-SNAPSHOT.jar
    lib/geode-rebalancer-1.8.0-SNAPSHOT.jar
    lib/geode-wan-1.8.0-SNAPSHOT.jar
```

**Comment by pivotal-jbarrett** [comment]
@metatype This latest commit will correct those missing geodes and locally the verify assembly test is 100% match. Check the CI integration test for pass and those issues should be resolved.

**Comment by metatype** [comment]
There's one more thing to fix:  the manifest classpath in `geode-dependencies.jar`.  Looks like it's missing commons-codec and lucene-analyzers-phonetic.

**Comment by pivotal-jbarrett** [comment]
Well crap. I’ll write another test. Anything else?

**Comment by metatype** [comment]
Not sure...geode-pulse, geode-web-api, and geode-web all use `classpath configurations.runtime` to build the war.  Might be worth double-checking that.

**Comment by pivotal-jbarrett** [comment]
@metatype this last push corrects all outstanding war, rar and other such assemblies.
